### PR TITLE
Issue 431: core + core_question changes: range filter

### PR DESCRIPTION
**Pull request by marcghaly**
This PR has changes for `range filter` only. It is for issue #431.

Cheers

### CODE REVIEW

File: lib/amd/src/local/filter/range.js [hunk 1]
```diff
@@ -0,0 +1,219 @@
...
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with Moodle.  If not, see <http://www.gnu.org/licenses/>.
+
+/**
+ * Range filter.
+ *
+ * @module     core/local/filter/range
+ * @copyright  2021 Ghaly Marc-Alexandre <marc-alexandreghaly@catalyst-ca.net>
```
Comment (safatshahin) [inline lib/amd/src/local/filter/range.js:20]: please fix the copyright

File: lib/amd/src/local/filter/range.js [hunk 2]
```diff
@@ -0,0 +1,219 @@
...
+        .catch();
+    }
+
+    /**
+     * Get the placeholder for range value one.
+     *
+     * @return {String} String
+     */
+    get placeholderOne() {
+        return 'First range value';
```
Comment (safatshahin) [inline lib/amd/src/local/filter/range.js:152]: use lang string instead

File: lib/amd/src/local/filter/range.js [hunk 3]
```diff
@@ -0,0 +1,219 @@
...
+    }
+
+    /**
+     * Get the placeholder for range value two.
+     *
+     * @return {String} String
+     */
+    get placeholderTwo() {
+        return 'Second range value';
```
Comment (safatshahin) [inline lib/amd/src/local/filter/range.js:161]: lang string here

File: lib/templates/filter_range.mustache [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+    GNU General Public License for more details.
+
+    You should have received a copy of the GNU General Public License
+    along with Moodle.  If not, see <http://www.gnu.org/licenses/>.
+}}
+{{!
+    @template core/filter_range
+
+    Template for the form containing one or more filter rows.
```
Comment (safatshahin) [inline lib/templates/filter_range.mustache:20]: add template explanation of the variables

File: lib/templates/local/filter/range.mustache [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+    GNU General Public License for more details.
+
+    You should have received a copy of the GNU General Public License
+    along with Moodle.  If not, see <http://www.gnu.org/licenses/>.
+}}
+{{!
+    @template core/local/filter/range
+
+    Moodle template for range filter.
+
```
Comment (safatshahin) [inline lib/templates/local/filter/range.mustache:21]: explain the variables

File: question/amd/src/filter.js [hunk 6]
```diff
@@ -113,7 +113,12 @@ export const init = (filterRegionId, defaultcourseid, defaultcategoryid,
 
             // Retrieve fitter info.
             for (const [key, value] of Object.entries(filterdata)) {
-                let filter = {'filtertype': key, 'jointype': value.jointype, 'values': value.values.toString()};
+                let filter = {
```
Comment (safatshahin) [inline question/amd/src/filter.js:116]: add this change to issue 431 PR
and remove from this commit

File: question/classes/external/qbank_filter.php [hunk 7]
```diff
@@ -61,6 +61,7 @@ public static function execute_parameters(): external_function_parameters {
                         'filtertype' => new external_value(PARAM_ALPHANUM, 'Filter type'),
                         'jointype' => new external_value(PARAM_INT, 'Join type'),
                         'values' => new external_value(PARAM_RAW, 'list of ids'),
+                        'rangetype' => new external_value(PARAM_INT, 'Range type', VALUE_OPTIONAL),
```
Comment (safatshahin) [inline question/classes/external/qbank_filter.php:64]: add this change to issue 431 PR
and remove from this commit

File: question/classes/external/qbank_filter.php [hunk 8]
```diff
@@ -170,7 +171,8 @@ public static function execute(
         foreach ($filters as $filter) {
             $params['filters'][$filter['filtertype']] = [
                 'jointype' => $filter['jointype'],
-                'values' => empty($filter['values']) ? [] : explode(',', $filter['values']),
+                'rangetype' => $filter['rangetype'] ?? null,
```
Comment (safatshahin) [inline question/classes/external/qbank_filter.php:174]: add this change to issue 431 PR
and remove from this commit

File: question/classes/external/qbank_filter.php [hunk 9]
```diff
@@ -170,7 +171,8 @@ public static function execute(
...
+                'values' => empty($filter['values'])  && $filter['values'] != 0 ? [] : explode(',', $filter['values']),
```
Comment (safatshahin) [inline question/classes/external/qbank_filter.php:175]: add this change to issue 431 PR
and remove from this commit
same for the rest

File: question/classes/local/bank/condition.php [hunk 10]
```diff
@@ -38,6 +38,18 @@ abstract class condition {
     /** @var int All of the following match */
     const JOINTYPE_ALL = 2;
 
+    /** @var int The default filter type (BETWEEN) */
```
Comment (safatshahin) [inline question/classes/local/bank/condition.php:41]: add this change to issue 431 PR
and remove from this commit

File: lib/amd/src/local/filter/range.js (hunk 1 above)
Comment (marcghaly) [reply lib/amd/src/local/filter/range.js:20]: Fixed

File: lib/amd/src/local/filter/range.js (hunk 2 above)
Comment (marcghaly) [reply lib/amd/src/local/filter/range.js:152]: Fixed

File: lib/amd/src/local/filter/range.js (hunk 3 above)
Comment (marcghaly) [reply lib/amd/src/local/filter/range.js:161]: Fixed

File: lib/templates/filter_range.mustache (hunk 4 above)
Comment (marcghaly) [reply lib/templates/filter_range.mustache:20]: Fixed

File: lib/templates/local/filter/range.mustache (hunk 5 above)
Comment (marcghaly) [reply lib/templates/local/filter/range.mustache:21]: Fixed

**Comment by marcghaly** [comment]
Conversation have been fixed and branch rebased over last `master_MDL-72321-api-changes`

File: lib/amd/src/local/filter/range.js [hunk 11]
```diff
@@ -0,0 +1,221 @@
...
+        this.initialValues = initialValues;
+        this.filterRange = filterRange;
+
+        this.setUpRangeUi('text');
+    }
+
+    /**
+     * Perform any tear-down for this filter type.
+     */
+    tearDown() {
```
[the remark this answers is not in the document]
Comment (marcghaly) [reply lib/amd/src/local/filter/range.js:56]: The purpose is the to trigger the clear filter action, it is also used in base filter

File: lang/en/moodle.php [hunk 12]
```diff
@@ -1725,6 +1725,8 @@
 $string['questionsinthequestionbank'] = 'Questions in the question bank';
 $string['queued'] = 'Queued';
 $string['quotausage'] = 'You have currently used {$a->used} of your {$a->total} limit.';
+$string['rangevalueone'] = 'First range value';
```
[the remark this answers is not in the document]
Comment (marcghaly) [reply lang/en/moodle.php]: Fixed

File: lang/en/moodle.php [hunk 13]
```diff
@@ -1725,6 +1725,8 @@
...
+$string['rangevaluetwo'] = 'Second range value';
```
[the remark this answers is not in the document]
Comment (marcghaly) [reply lang/en/moodle.php]: Fixed

**Comment by tuanngocnguyen** [comment]
Rebased over latest master_MDL-72321-api-changes (with phpunit test) and pushed

**Comment by tuanngocnguyen** [comment]
This MR looks good to me.

**Pull request merged**